### PR TITLE
feat(fp6): MX-FP6 (W6A8) dense + MoE serving stack for SM12.0

### PR DESCRIPTION
**Pull request by phaelon74**
## Summary

This PR adds a production FP6 serving path for SM12.0 (Blackwell workstation/consumer:
RTX PRO 6000, GeForce RTX 50) to SparkInfer. The goal was to follow the library's
existing lead: FP6 slots in beside MXFP8, NVFP4 and W4A8 as a peer quant mode — same
`plan`/`bind`/`run` facades, same op registration, same checkpoint conventions — reusing
the proven block-scaled GEMM and MoE machinery and adding only what FP6 genuinely
requires.

One squashed commit: **79 files, +19,575 / −280**, rebased onto current `master`.

Full design record: `docs/FP6_in_SparkInfer(B12X).md`. Reference docs:
`docs/mxfp6-w6a8.md` (format/kernels), `docs/fp6-user-guide.md` (making & serving
quants), `docs/fp6-results.md` (KLD + performance), `docs/mxfp6-vllm-integration.md`
(vLLM plugin).

## Format

- **On disk: weight-only W6A16.** Packed MX-FP6 E2M3 codes (4 values in 3 bytes along
  K), MSE-refined UE8M0 K/32 block scales, per-tensor (dense) / per-expert (MoE) f32
  global scales, ModelOpt-mirror safetensors schema. Calibration-free — a quant is a
  pure weight transform, no dataset needed. Safetensors only, no pickle anywhere.
- **At runtime: W6A8.** Activations are quantized every forward to FP8 E4M3 with UE8M0
  K/32 scales. E4M3 activations roughly halve the runtime KLD slice vs 6-bit
  activations, cost nothing at the tensor core (the `mxf8f6f4` MMA consumes FP8 A ×
  FP6 B natively at full rate), and reuse the `w4a8_mx` activation machinery unchanged.
  True-W6A6 remains selectable at export time for A/B.

## Kernels

- **Dense GEMM**: FP6 rides the existing MXFP8 block-scaled pipeline as `uint8`
  byte-containers — the PTX-mandated operand format for `e2m3`/`e3m2` under
  `.kind::mxf8f6f4` — with the per-K-block MMA emitted as inline PTX with explicit
  FP6/FP8 type qualifiers (`_lib/dense_gemm_mxfp6.py`). Packed-B streaming (3:4 form in
  HBM, expanded in smem) saves 25% of B-side HBM traffic on large-N layers.
- **Activation quantizers**: a TMA-based large-M quantizer for prefill, and a small-M
  (m ≤ 16) decode/MTP quantizer with the full per-row scale recipe fused in-kernel
  (row amax, correctly-rounded scale, bf16 pre-scale, quant, inverse scale + alpha for
  the epilogue) — zero host-side launches on the decode hot path.
- **MoE**: `quant_mode="w6a8_mx"` (`source_format="mxfp6_e2m3"`), following the
  `w4a8_mx` shape: per-expert alphas, unified dynamic persistent-grid scheduling,
  optional bit-deterministic combine (`SPARKINFER_DYNAMIC_DETERMINISTIC_OUTPUT=1`,
  scoring only).
- **Small-N BF16 GEMV** (`gemm.bf16_gemv`) for narrow uncovered projections
  (e.g. GDN `in_proj_ba`, N ≤ 1024).

## Bit-exact determinism

The accuracy-scoring contract this PR holds: **KLD is bit-identical every run and
across every equivalent execution path** (fused vs host-chain, small-M vs large-M,
decode vs prefill row).

- Per-row activation global scales make each row's quantization independent of batch
  composition (chunked-prefill safe; an m=1 decode row matches its prefill twin
  bit-for-bit).
- All scale derivations are correctly rounded: a new `div_rn_f32` PTX intrinsic
  in-kernel, and f64-divide-then-cast on the host (torch's CUDA f32 division is not
  always correctly rounded — e.g. `200704 / 2.625` lands 1 ulp high).

## Measured results (RTX PRO 6000, sm_120)

**Accuracy** — Wikitext-2 KLD vs BF16, ctx 2048 / stride 512, eager + deterministic
combine:

| Model | KLD | Reproducibility |
|---|---|---|
| Qwen3.6-27B dense (W6A8) | 0.034599 | bit-identical across 3 runs incl. host-chain fallback |
| Qwen3.6-35B-A3B MoE (W6A8) | 0.015388 | bit-identical across runs (FP8 band) |

**Performance** — single-stream serving with MTP, full 1k–32k context sweeps in
`docs/fp6-results.md`:

| Model | TP=1 TPOT | TP=2 TPOT |
|---|---|---|
| Qwen3.6-27B dense | 9.79 ms/token | 7.49 ms/token |
| Qwen3.6-35B-A3B MoE | 4.28 ms/token | 3.94 ms/token |

## Integration and tooling

- `sparkinfer/integration/vllm/`: a thin plugin registered via vLLM's
  `general_plugins` entry point plus a framework-agnostic `fp6_serving` layer,
  mirroring how the NVFP4 path is wired. MoE plans/scratch are deduplicated by
  geometry across layers (keeps the footprint flat under vLLM's CUDA-graph memory
  estimator).
- `scripts/quantize_model_fp6.py` / `quantize_moe_fp6.py` /
  `dequantize_fp6_to_bf16.py`: HF checkpoint → FP6 safetensors export with
  architecture discovery and dry-run/inspect modes.

## Scope and caveats

- **SM12.0 first and foremost.** All development and validation on RTX PRO 6000.
  SM12.1 shares the instruction set and is expected to work but is unvalidated;
  datacenter Blackwell (tcgen05/TMEM path) is out of scope; pre-Blackwell has no
  `mxf8f6f4` MMA and cannot be supported as designed.
- The deterministic MoE combine is a scoring tool (1.3–4.4× slower combine), not a
  serving default.
- vLLM TP>1 on sm_120 requires `--disable-custom-all-reduce` (vLLM's custom
  all-reduce faults during CUDA-graph capture; NCCL fallback costs ~1–3%).
- **Pre-existing test failures**: 7 regression-test failures on non-FP6 paths exist on
  `master` before this branch and are unrelated to this PR (list available on request).

## Test plan

- [x] Unit tests: quantizer-vs-TMA bit-equality, fused-vs-host per-row bit-equality
  (m ∈ {1,3,5,16} × all formats), end-to-end small-M vs M=128 row equality, compile
  guards, dense pipeline roundtrip/equivalence — all passing on sm_120.
- [x] KLD scoring: dense and MoE, repeated runs, bit-identical (values above).
- [x] Serving: dense 27B and MoE 35B-A3B under vLLM (nightly) with MTP, TP=1 and TP=2,
  1k–32k context benchmark sweeps.
- [x] Quantization tooling: fresh dense + MoE exports, load-and-serve verified.

## Summary by CodeRabbit

* **New Features**
  * Added MX-FP6 (W6A6/W6A8) dense + MoE serving support, including the W6A8-MX recipe and an FP6 dense custom op.
  * Added a BF16 small‑N GEMV accelerator and an opt-in FP6 MoE decode micro mode.
* **Documentation**
  * Added FP6 user guide, results, and vLLM integration docs (plus engineering records).
* **Benchmarks**
  * Added BF16→FP6, dense GEMM, MoE, decode micro, and serving TPS benchmarks.
* **Tests**
  * Added extensive CUDA correctness, determinism, quantization/packing, and checkpoint IO coverage.
* **Chores**
  * Updated git ignore for local validation notes.

### CODE REVIEW

File: benchmarks/bench_bf16_to_fp6_tma.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    cols_pad_sf = ((k // 32 + 3) // 4) * 4
+    inp = (
+        bf16
+        if rows_padded == m and bf16.is_contiguous()
+        else torch.zeros((rows_padded, k), dtype=torch.bfloat16, device=dev)
+    )
+    if rows_padded != m:
+        inp[:m].copy_(bf16)
+    out = allocate_bf16_to_fp6_tma_outputs(m, k, device=dev)
+    compiled = compile_bf16_to_fp6_tma(rows_padded, k)
```
[the remark this answers is not in the document]
Comment (phaelon74) [reply benchmarks/bench_bf16_to_fp6_tma.py]: Resolved in https://github.com/local-inference-lab/sparkinfer/pull/78/changes/28dd0a9cf836d693a556a7d31269bab5ccb1d704

File: benchmarks/benchmark_moe_fp6.py [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+        starts[i].record()
+        replay()
+        ends[i].record()
+    torch.cuda.synchronize()
+    times = [s.elapsed_time(e) for s, e in zip(starts, ends)]
+    med = statistics.median(times)
+    print(
+        f"W6A8 MX-FP6 MoE synthetic m={m} k={k} n={n} E={e} topk={topk}: "
+        f"{fmt_us(times)}  (median {med:.3f} ms)"
+    )
```
[the remark this answers is not in the document]
Comment (phaelon74) [reply benchmarks/benchmark_moe_fp6.py:159]: Resolved in https://github.com/local-inference-lab/sparkinfer/pull/78/changes/28dd0a9cf836d693a556a7d31269bab5ccb1d704

File: scripts/capture_hidden_states.py [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+        extra = {"device_map": "auto"}
+        move_after = False
+    except ImportError:
+        # device_map requires accelerate; load on CPU and move instead.
+        extra = {}
+        move_after = torch.cuda.is_available()
+    print(f"[capture] loading model (bf16, {extra or 'cpu->cuda'}) ...")
+    model = AutoModelForCausalLM.from_pretrained(
+        args.model, dtype=torch.bfloat16, trust_remote_code=True, **extra
+    )
```
[the remark this answers is not in the document]
Comment (phaelon74) [reply scripts/capture_hidden_states.py]: Resolved in https://github.com/local-inference-lab/sparkinfer/pull/78/changes/28dd0a9cf836d693a556a7d31269bab5ccb1d704

File: sparkinfer/_lib/fp6.py [hunk 4]
```diff
@@ -0,0 +1,1823 @@
...
+    """
+    r0 = row >> Int32(7)  # row // 128
+    r1 = (row >> Int32(5)) & Int32(3)  # (row % 128) // 32
+    r2 = row & Int32(31)  # row % 32
+    c0 = block >> Int32(2)  # block // 4
+    c1 = block & Int32(3)  # block % 4
+    return (
+        ((((r0 * cols_padded_div4 + c0) * Int32(32) + r2) * Int32(4) + r1) * Int32(4))
+        + c1
+    )
```
[the remark this answers is not in the document]
Comment (phaelon74) [reply sparkinfer/_lib/fp6.py:206]: Resolved in https://github.com/local-inference-lab/sparkinfer/pull/78/changes/28dd0a9cf836d693a556a7d31269bab5ccb1d704

File: sparkinfer/gemm/bf16_gemv/_kernel.py [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+                s = acc[r]
+                for wv, xv in ((w0, x0), (w1, x1), (w2, x2), (w3, x3)):
+                    s = s + u32_as_f32(wv << Uint32(16)) * u32_as_f32(
+                        xv << Uint32(16)
+                    )
+                    s = s + u32_as_f32(wv & Uint32(0xFFFF0000)) * u32_as_f32(
+                        xv & Uint32(0xFFFF0000)
+                    )
+                acc[r] = s
+            i += Int32(_THREADS)
```
[the remark this answers is not in the document]
Comment (phaelon74) [reply sparkinfer/gemm/bf16_gemv/_kernel.py]: Resolved in https://github.com/local-inference-lab/sparkinfer/pull/78/changes/28dd0a9cf836d693a556a7d31269bab5ccb1d704

File: sparkinfer/quantization/mxfp6/bf16_to_fp6_tma.py [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+                    #   tile base = (mt*k_tiles + kt) * 512
+                    #   within-tile = (row%32)*16 + (row//32)*4 + sf_block
+                    outer_m_idx = row % Int32(32)
+                    inner_m_idx = row // Int32(32)
+                    sf_offset = (
+                        (mt * k_tiles + kt) * Int32(512)
+                        + outer_m_idx * Int32(16)
+                        + inner_m_idx * Int32(4)
+                        + sf_block
+                    )
```
[the remark this answers is not in the document]
Comment (phaelon74) [reply sparkinfer/quantization/mxfp6/bf16_to_fp6_tma.py:331]: Resolved in https://github.com/local-inference-lab/sparkinfer/pull/78/changes/28dd0a9cf836d693a556a7d31269bab5ccb1d704

File: sparkinfer/quantization/mxfp6/fp6_checkpoint.py [hunk 7]
```diff
@@ -0,0 +1,512 @@
...
+    use_lo = (mse_lo < mse_hi) & nonzero
+
+    codes = torch.where(use_lo, codes_lo, codes_hi)
+    codes = torch.where(nonzero, codes, torch.zeros_like(codes))
+    e_chosen = torch.where(use_lo, e - 1.0, e)
+    ue = (e_chosen.squeeze(-1).to(torch.int32) + 127).clamp(0, 255).to(torch.uint8)
+    ue = torch.where(amax.squeeze(-1) <= 0, torch.zeros_like(ue), ue)
+
+    packed = pack_fp6_codes_tensor(codes.reshape(rows, k))
+    return packed, ue
```
[the remark this answers is not in the document]
Comment (phaelon74) [reply sparkinfer/quantization/mxfp6/fp6_checkpoint.py:422]: Resolved in https://github.com/local-inference-lab/sparkinfer/pull/78/changes/28dd0a9cf836d693a556a7d31269bab5ccb1d704

File: sparkinfer/quantization/mxfp6/fp6_moe_weights.py [hunk 8]
```diff
@@ -0,0 +1,270 @@
...
+
+    tensors: dict[str, torch.Tensor] = {}
+    # historical schema id; do not rename (existing artifacts carry it)
+    metadata = {"__format__": "b12x_fp6_moe_weights_v1"}
+    for key, value in asdict(weights).items():
+        if isinstance(value, torch.Tensor):
+            tensors[key] = value.detach().cpu().contiguous()
+        else:
+            metadata[key] = json.dumps(value)
+    save_file(tensors, path, metadata=metadata)
```
[the remark this answers is not in the document]
Comment (phaelon74) [reply sparkinfer/quantization/mxfp6/fp6_moe_weights.py]: Resolved in https://github.com/local-inference-lab/sparkinfer/pull/78/changes/28dd0a9cf836d693a556a7d31269bab5ccb1d704

File: tests/quantization/test_bf16_to_fp6_tma_setup.py [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+    compile_bf16_to_fp6_tma,
+)
+from sparkinfer.quantization.mxfp6.bf16_to_fp6_tma import TestKernel
+
+
+def test_bf16_to_fp6_tma_kernel_import() -> None:
+    assert TestKernel is not None
+    k = TestKernel()
+    assert k.tile_shape_mnk == (128, 128, 128)
+    assert k.threads_per_cta == 160
```
[the remark this answers is not in the document]
Comment (phaelon74) [reply tests/quantization/test_bf16_to_fp6_tma_setup.py]: Resolved in https://github.com/local-inference-lab/sparkinfer/pull/78/changes/28dd0a9cf836d693a556a7d31269bab5ccb1d704

File: scripts/bench_fp6_micro.py [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+            os.environ["SPARKINFER_ENABLE_FP6_MICRO"] = "1"
+            run = make_runner(m, x, topk_ids, topk_weights)
+            micro_out = run().clone()
+            t_micro = _bench_once(run, args.iters, args.warmup)
+
+            cos = torch.nn.functional.cosine_similarity(
+                micro_out.reshape(-1).float(), dyn_out.reshape(-1).float(), dim=0
+            ).item()
+            speedup = t_dyn / t_micro if t_micro > 0 else float("nan")
+            print(f"{m:>4} {t_dyn:>12.4f} {t_micro:>12.4f} {speedup:>9.2f} {cos:>9.4f}")
```
[the remark this answers is not in the document]
Comment (phaelon74) [reply scripts/bench_fp6_micro.py:200]: Ddopted as suggested, plus an independent BF16 ground-truth gate on the dynamic path so a shared-bug failure mode is also covered.

Covered in https://github.com/local-inference-lab/sparkinfer/pull/78/changes/0a36afb5ec07123a2f977ba56190ab91f8380bf0
